python/ray/serve/controller.py

```diff
@@ -408,25 +406,79 @@ def deploy_app(
             self.deployment_timestamp = time.time()
 
         config_dict = config.dict(exclude_unset=True)
+
+        # Compare new config options with old config options and set versions of new deployments
+        config_checkpoint = self.kv_store.get(CONFIG_CHECKPOINT_KEY)
+        if config_checkpoint is not None:
+            _, old_config_dict, version_dict = pickle.loads(config_checkpoint)
+            version_dict = self.version_config(
+                config_dict, old_config_dict, version_dict
+            )
+        else:
+            version_dict = self.version_config(config_dict, {}, {})
+
         self.kv_store.put(
             CONFIG_CHECKPOINT_KEY,
-            pickle.dumps((self.deployment_timestamp, config_dict)),
+            pickle.dumps((self.deployment_timestamp, config_dict, version_dict)),
         )
 
+        deployment_override_options = config_dict.get("deployments", [])
+
         if self.config_deployment_request_ref is not None:
             ray.cancel(self.config_deployment_request_ref)
             logger.info(
                 "Received new config deployment request. Cancelling "
                 "previous request."
             )
 
-        deployment_override_options = config.dict(
-            by_alias=True, exclude_unset=True
-        ).get("deployments", [])
-
         self.config_deployment_request_ref = run_graph.options(
             runtime_env=config.runtime_env
-        ).remote(config.import_path, config.runtime_env, deployment_override_options)
+        ).remote(
+            config.import_path,
+            config.runtime_env,
+            deployment_override_options,
+            version_dict,
+        )
+
+    def version_config(
+        self, new_config: Dict, old_config: Dict, version_dict: Dict
+    ) -> Dict[str, str]:
+        """
+        Compares an old config and a new config, dicts that follow ServeApplicationSchema.
+
+        For each deployment, if any field other than num_replicas, user_config,
+        and autoscaling_config changed, bump the version for that deployment
+        to a new random version.
+        """
```

Comment:
I'd recommend separating out the specific field names in this docstring to make it easier to update:

```
When a deployment's options change, its version should generally change,
so old replicas are torn down. The only options which can be updated
without tearing down replicas (i.e. changing the version) are:

* num_replicas
* user_config
* autoscaling_config

This function determines whether each deployment's version
should be changed based on the updated options.
```